test/unit_tests/braket/experimental/autoqasm/test_transpiler.py

```diff
@@ -71,50 +68,74 @@ def bell(q0: int, q1: int):
         h(q0)
         cnot(q0, q1)
 
-    @aq.main
-    def bell_decorated(q0: int, q1: int):
-        bell(q0, q1)
+    expected_partial = """OPENQASM 3.0;
+input int[32] q1;
+qubit[2] __qubits__;
+h __qubits__[1];
+cnot __qubits__[1], __qubits__[q1];"""
 
-    expected = """OPENQASM 3.0;
-qubit[4] __qubits__;
+    expected_no_arg_partial = """OPENQASM 3.0;
+qubit[2] __qubits__;
 h __qubits__[1];
 cnot __qubits__[1], __qubits__[3];"""
 
-    bell_partial = aq.main(functools.partial(bell, 1))
-    assert bell_partial(3).to_ir() == expected
+    bell_partial = aq.main(num_qubits=2)(functools.partial(bell, 1))
```

Comment:
Why do we have to supply `num_qubits` on this input now? Maybe it's a rare enough use case that we don't have to worry about it. @rmshaffer, thoughts?

Reply:
Here we need `num_qubits` because the program is built with variable qubit indices in use. For the case below, where we pass in a partial fully binding the qubit args, `num_qubits` can be omitted. This seems correct to me.

It's worth noting that this function previously always built the Program with all variables bound, even though the `aq.main` call was on varying partials. Now that building happens during the decorator call, incomplete partials will have free parameters.

We could test partials on a meta-programming approach that would more closely match the end result of the previous test, but I think this approach is more fitting, as the essence of this test was asserting the behavior of decorating partial functions.

Reply:
Agreed, this new behavior seems reasonable and the test makes sense to me.